pkg/pubsub/dapr/dapr.go

```diff
@@ -0,0 +1,114 @@
+package dapr
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/dapr/go-sdk/client"
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub/common"
+	"github.com/open-policy-agent/gatekeeper/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeClient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	Name = "dapr"
+)
+
+type Endpoint struct {
+	// Name of the component to be used for pub sub messaging
+	Component string `json:"component"`
+
+	// Sources from which the messages are allowed to be published using this client, e.g Audit, Validation
+	Source []common.Source `json:"source"`
+}
+
+type ClientConfig struct {
+	// Enable, Disable batching per tool
+	EnableBatching bool `json:"enableBatching"`
+
+	// Different endpoints to publish messages
+	Endpoints []Endpoint `json:"endpoints"`
+}
+
+type Client struct {
+	client          client.Client
+	pubSubComponent string
+	source          []common.Source
+}
+
+// Dapr represents driver for interacting with pub sub using dapr.
+type Dapr struct {
+	// Array of clients to talk to different endpoints
+	client []Client
+
+	// Name of the pubsub tool
+	name string
+
+	// Enable, Disable batching per tool
+	batchingEnabled bool
+}
+
+func (r *Dapr) Publish(data interface{}, topic string, source common.Source) error {
+	jsonData, err := json.Marshal(data)
+	if err != nil {
+		return fmt.Errorf("error marshaling data: %w", err)
+	}
+
+	for _, c := range r.client {
+		for _, src := range c.source {
+			if src == source {
+				err = c.client.PublishEvent(context.Background(), c.pubSubComponent, topic, jsonData)
+				if err != nil {
+					return fmt.Errorf("error publishing message to dapr: %w", err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *Dapr) PublishBatch(data interface{}, topic string, source common.Source) error {
+	return nil
+}
+
+func (r *Dapr) NewClient(ctx context.Context, k8sClient kubeClient.Client) error {
+	var err error
+	config := corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Name: Name, Namespace: util.GetNamespace()}}
+	err = k8sClient.Get(ctx, kubeClient.ObjectKey{Name: Name, Namespace: util.GetNamespace()}, &config)
```

Comment:
Sample config input

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: dapr
  namespace: gatekeeper-system
data:
  dapr.json: |
    {
      "enableBatching": true,  
      "endpoints": [
        {
          "component": "redis",
          "source": ["Audit"]
        },
        {
          "component": "rabbitMQ",
          "source": ["Audit"]
        }
      ]
    }
```

Reply:
Where do the secrets go?

Reply:
That will depend on per tool basis right? to initialize dapr client, there is no need to pass the secret in

Reply:
Doesn't dapr need a secret to authenticate the Dapr connection?

https://docs.dapr.io/concepts/security-concept/

Reply:
It looks like Dapr expects a sidecar. I'd expect this to be the exception rather than the rule. We should not overfit on dapr.

Dapr is not a good fit for `gator`, nor is it likely to be useful for platform providers, since it requires a singleton system enabled on the cluster, which providers may not be able to ensure exists (or users may want to manage for themselves)

Reply:
According to [this section](https://docs.dapr.io/concepts/security-concept/#secure-dapr-to-control-plane-communication), I understand that dapr takes care of certs inherently while initializing sidecar, since there is no way to pass credentials in `NewClient` or `Publish`. So, for at least dapr's case, we might not need to pass secret.

For other tools, It will depend on the driver. How about an additional parameter for `NewClient` alongside config string to pass in secret data as well, and then unmarshal secret in client as per tool requirement? would that work? do you think there are any shortcomings on implementing this?